doc/holochain_101/src/writing_development_kit.md

```diff
@@ -1 +1,100 @@
+- hc_send
+- hc_start_bundle
+- hc_close_bundle
+
+There is a special additional one called `hc_init_globals` which we will discuss further.
+
+The Development Kit should implement and export one function per each in the above list. The function should be called the same as its native form, but without the prefix. E.g. `hc_update_agent` should be called `update_agent` or `updateAgent`. That function should internally call the native function and handle the additional complexity around that.
+
+In order to call these "external" functions, you will need to import them and provide their signature, but in a WASM import compatible way. In Rust, for example, this is simply:
+```rust
+extern {
+  fn hc_commit_entry(encoded_allocation_of_input: u32) -> u32;
+}
+```
+
+TODO: define or link to meaningful function signatures
+
+### Working with WASM Memory
+
+The goal of the Development Kit is to expose a meaningful and easy to use version of the API functions, with meaningful arguments and return values. There is a bit of flexibility around how this is done, as coding languages differ. However, the internal process will be similar in nature. Here it is, generalized:
+1. declare, or use a passed, single page 64 KiB memory stack
+2. join whatever inputs are given into a single serializable struct
+3. serialize the given data structure as an array of bytes
+4. determine byte array length
+5. ensure it is not oversized for the stack
+6. allocate the memory
+7. write the byte array to memory
+8. create an allocation pointer for the memory  
+  a. use a 16 bit integer for the pointers `offset`  
+  b. use a 16 bit integer for the pointers `length`
+9. join the pointers into a single 32 bit integer  
+  a. high bits are `offset`  
+  b. low bits are `length`
+10. call the native function with that 32 bit integer and assign the result to another 32 bit integer  
+  a. e.g. `encoded_alloc_of_result = hc_commit_entry(encoded_alloc_of_input)`
+11. deconstruct that 32 bit integer into two variables  
+  a. use a 16 bit integer for the pointers `offset`  
+  b. use a 16 bit integer for the pointers `length`
+12. read string data from memory at the `offset` address
+13. deallocate the memory
+14. deserialize the string to JSON if JSON is expected
+
+That looks like a lot of steps, but most of this code can be shared for the various functions throughout the Development Kit, leaving implementations to be as little as 5 lines long. Basically, the process inverts at the point of the native function call.
+
+#### WASM Single Page Stack
+
+TODO
+
+### App Globals
+
+When writing Zome code, it is common to need to reference aspects of the context it runs in, such as the active user/agent, or the DNA hash of the app. Holochain exposes certain values through to the Zome, though it does so natively by way of the `hc_init_globals` function mentioned. Again, to simplify the developer experience, it is valuable to expose these values as constants through to the developer.
```

Comment:
Awkward last sentence. Maybe "Taking care to expose these values as constants will simplify the developer experience" instead.